### PR TITLE
Fix/roles revamp #1772

### PR DESCRIPTION
**Pull request by alanarvelo**
Hi all,

This is a first pass at the Issue *Proposal to revamp Roles [#1772](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/1772)*. Merely an implementation of the `RoleChecker.sol` contract, whose intellectual author is @nventuro. Actual repo functionality was not altered. Made a PR to facilitate the discussion over the implementation details.

A few questions:
1. Is this implementation of `RoleChecker.sol` accurate and viable?
2. What should happen to the Role Contracts defined under `/access/roles` e.g. CapperRole, MinterRole, PauserRole, etc.? Should they be refactored to abide by the new paradigm as depicted by @nventuro in issue [#1772](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/1772), or should those contracts be deleted and all roles & functionality be defined where they are needed/used?
3. Is there a particular reason for the `roleIds` — hashes of address + counter — to be type `uint256` instead of the return type of keccak256 `bytes32`?

TO-DO:
1. Refactor all Roles in the repo to use new Roles paradigm
2. Verify all tests pass and possibly new ones

Looking forward to continue working on this!

### CODE REVIEW

File: contracts/access/RoleChecker.sol [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    event RoleRemoved(address indexed account, uint256 roleId);
+
+    modifier onlyRole(uint256 _roleId) {
+        require(hasRole(msg.sender, _roleId), "Caller does not have the permissions to call this function");
+        _;
+    }
+
+    function roleExists(uint256 _roleId) internal view returns (bool) { return roles[_roleId].exists; }
+
+    function newRole() internal returns (uint256) {
```
Comment (nventuro) [inline contracts/access/RoleChecker.sol]: We may want to add a string here, to help with debugging

File: contracts/access/RoleChecker.sol [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    event RoleAdded(address indexed account, uint256 roleId);
+    event RoleRemoved(address indexed account, uint256 roleId);
+
+    modifier onlyRole(uint256 _roleId) {
+        require(hasRole(msg.sender, _roleId), "Caller does not have the permissions to call this function");
+        _;
+    }
+
+    function roleExists(uint256 _roleId) internal view returns (bool) { return roles[_roleId].exists; }
```
Comment (nventuro) [inline contracts/access/RoleChecker.sol]: Interesting, I had not considered this! Can you think of use cases where someone might want to check if a role exists?

File: contracts/access/Roles.sol [hunk 3]
```diff
@@ -31,6 +32,7 @@ library Roles {
      */
     function has(Role storage role, address account) internal view returns (bool) {
         require(account != address(0), "Roles: account is the zero address");
+        // require(role.exists, "Roles: role does not exists");git
```
Comment (nventuro) [inline contracts/access/Roles.sol]: The fact that this check is not really necessary makes me think that perhaps having `.exists` is not the best way to go: it is possible to misuse the `Roles` struct by forgetting to set this field.

**Comment by nventuro** [comment]
Hello @alanarvelo! Sorry for taking so long to review this, I try to give feedback as early as possible and really dropped the ball here.

This is quite a big feature that we want to get right, so progress may not be super fast :)

>What should happen to the Role Contracts defined under /access/roles e.g. CapperRole, MinterRole, PauserRole, etc.? Should they be refactored to abide by the new paradigm as depicted by @nventuro in issue #1772, or should those contracts be deleted and all roles & functionality be defined where they are needed/used?

We should definitely keep the contracts and the current interface (see our [API stability guideline](https://docs.openzeppelin.com/contracts/2.x/api-stability). It feels like it should be possible to do a light rewrite using this new paradigm while keeping the external interface intact, but I guess we won't know until we try :)

>Is there a particular reason for the roleIds — hashes of address + counter — to be type uint256 
instead of the return type of keccak256 bytes32?

No, I simply tried to convey the fact that they are 256-bit ids. If `bytes32` is a more convenient type than `uint256`, then we should use it.

**Comment by nventuro** [comment]
We're definitely interested in getting this merged! If @alanarvelo is unable to continue the PR, is there anyone else that might take up development?

**Comment by frangio** [comment]
Putting this on hold until we have time to dedicate to it.

**Comment by nventuro** [comment]
This PR was superceded by #2112 - sorry to have kept it open for so long.

**Pull request closed without merging**